packages/eslint-plugin/tests/rules/restrict-template-expressions.test.ts

```diff
@@ -227,6 +227,44 @@ ruleTester.run('restrict-template-expressions', rule, {
         }
       `,
     },
+    // allowNever
+    {
+      options: [{ allowNever: true }],
+      code: `
+        const arg = 'hello';
+        const msg = typeof arg === 'string' ? arg : \`arg = \${arg}\`;
+      `,
+    },
+    {
+      options: [{ allowNever: true }],
+      code: `
+        function test(arg: 'one' | 'two') {
+          switch (arg) {
+            case 'one':
+              return 1;
+            case 'two':
+              return 2;
+            default:
+              throw new Error(\`Unrecognised arg: \${arg}\`);
+          }
+        }
+      `,
+    },
+    {
+      options: [{ allowNever: true }],
+      code: `
+        // more variants may be added to Foo in the future
+        type Foo = { type: 'a'; value: number };
+        
+        function checkFoosAreMatching(foo1: Foo, foo2: Foo) {
+          if (foo1.type !== foo2.type) {
+            // since Foo currently only has one variant, this code is never run, and \`foo1.type\` has type \`never\`.
+            throw new Error(\`expected \${foo1.type}, found \${foo2.type}\`);
+            // ^ Invalid type "never" of template literal expression.
+          }
+        }
+      `,
+    },
```

Comment:
I don't see any existing tests for the `never` type _without_ the `allowNever` option. Unless I'm missing them (am I?), could you add one in? We should make sure the new logic only applies when the option is enabled.

Reply:
Good idea. Have added a check that the rule still fires when allowNever is disabled.

Reply:
Awesome, thanks! Whenever you feel this is ready for re-review, you can re-request my review.

Btw we can ignore the failing website tests. #6508 😢